tests/FSharp.Core.UnitTests/FSharp.Core/PrimTypes.fs

```diff
@@ -905,32 +987,100 @@ module internal RangeTestsHelpers =
-    [<Fact>] member _.``Range.UInt64`` () = RangeTestsHelpers.unsigned System.UInt64.MinValue System.UInt64.MaxValue
+module RangeTests =
+    module BuildTime =
+        [<Fact>]
+        let ``Range.SByte``  () = RangeTestsHelpers.signed   System.SByte.MinValue  System.SByte.MaxValue
 
-    [<Fact>]
-    member _.``Range.IntPtr`` () =
-        if System.IntPtr.Size >= 4 then RangeTestsHelpers.signed (System.IntPtr System.Int32.MinValue) (System.IntPtr System.Int32.MaxValue)
-        if System.IntPtr.Size >= 8 then RangeTestsHelpers.signed (System.IntPtr System.Int64.MinValue) (System.IntPtr System.Int64.MaxValue)
-        
-    [<Fact>]
-    member _.``Range.UIntPtr`` () =
-        if System.UIntPtr.Size >= 4 then RangeTestsHelpers.unsigned (System.UIntPtr System.UInt32.MinValue) (System.UIntPtr System.UInt32.MaxValue)
-        if System.UIntPtr.Size >= 8 then RangeTestsHelpers.unsigned (System.UIntPtr System.UInt64.MinValue) (System.UIntPtr System.UInt64.MaxValue)
-        
+        [<Fact>]
+        let ``Range.Byte``   () = RangeTestsHelpers.unsigned System.Byte.MinValue   System.Byte.MaxValue
+
+        // Note: the IEnumerable<char> range iterator doesn't currently pass these tests. Should it?
+        //[<Fact>]
+        //let ``Range.Char``   () = RangeTestsHelpers.unsigned System.Char.MinValue   System.Char.MaxValue
+
+        [<Fact>]
+        let ``Range.Int16``  () = RangeTestsHelpers.signed   System.Int16.MinValue  System.Int16.MaxValue
+
+        [<Fact>]
+        let ``Range.UInt16`` () = RangeTestsHelpers.unsigned System.UInt16.MinValue System.UInt16.MaxValue
+
+        [<Fact>]
+        let ``Range.Int32``  () = RangeTestsHelpers.signed   System.Int32.MinValue  System.Int32.MaxValue
+
+        [<Fact>]
+        let ``Range.UInt32`` () = RangeTestsHelpers.unsigned System.UInt32.MinValue System.UInt32.MaxValue
+
+        [<Fact>]
+        let ``Range.Int64``  () = RangeTestsHelpers.signed   System.Int64.MinValue  System.Int64.MaxValue
+
+        [<Fact>]
+        let ``Range.UInt64`` () = RangeTestsHelpers.unsigned System.UInt64.MinValue System.UInt64.MaxValue
+
+        [<Fact>]
+        let ``Range.IntPtr`` () =
+            // 0x80000000n is negative on x86, but would be positive on x64.
+            if System.IntPtr.Size = 4 then RangeTestsHelpers.signed 0x80000000n         0x7fffffffn
+            if System.IntPtr.Size = 8 then RangeTestsHelpers.signed 0x8000000000000000n 0x7fffffffffffffffn
+            
+        [<Fact>]
+        let ``Range.UIntPtr`` () =
+            if System.UIntPtr.Size >= 4 then RangeTestsHelpers.unsigned 0x0un           0xffffffffun
+            if System.UIntPtr.Size >= 8 then RangeTestsHelpers.unsigned 0x0un           0xffffffffffffffffun
+
+    module RunTime =
+        [<Theory; InlineData(System.SByte.MinValue, System.SByte.MaxValue)>]
+        let ``Range.SByte``  (min0: sbyte) (max0: sbyte) = RangeTestsHelpers.signed   min0 max0
```

Comment:
I haven't committed to this place in the code yet so I might be terrible misunderstanding something - what would be the difference between BuildTime.Range.SByte test and RunTime.Range.SByte test? Why have both?

Reply:
The `RangeTestsHelpers.signed` and `RangeTestsHelpers.unsigned` helpers are marked `inline` (presumably originally to avoid duplicating them for each type).

This means, though, that when literal values are passed for `min0` and `max0` — as in the "build-time" tests — `start`, `step`, and `finish` actually show up in the TAST as `Expr.Const`s, and so the "build-time" pathway in the compiler is exercised, i.e., the iteration count is computed at build-time instead of the compiler emitting code to compute it at runtime.

The "run-time" tests intentionally do _not_ directly pass in literals for `min0` and `max0`, because otherwise they would be inlined as constants; they are meant to exercise the "runtime" pathway, i.e., the code emitted to compute the count at runtime.

Reply:
Alright, that's fine, probably we can add a comment for this :)

Reply:
574595a